### PR TITLE
Add hotplug and hotunplug cases into i6300esb watchdog device

### PR DESCRIPTION
**Pull request by iccaszhulili**
Now, only support hotplug and hotunplug operation to i6300esb
watchdog device.

Signed-off-by: Lily Zhu <lizhu@redhat.com>

### CODE REVIEW

**Comment by iccaszhulili** [comment]
# avocado run --vt-type libvirt watchdog
JOB ID     : 5d52eeea7496e284fbfced3dd2cc5a8f99a92937
JOB LOG    : /root/avocado/job-results/job-2018-07-05T22.29-5d52eee/job.log
 (01/31) type_specific.io-github-autotest-libvirt.watchdog.action_shutdown.model_i6300esb.model_test: /^[[PASS (57.07 s)
 (02/31) type_specific.io-github-autotest-libvirt.watchdog.action_shutdown.model_i6300esb.hotplug_test: PASS (50.92 s)
 (03/31) type_specific.io-github-autotest-libvirt.watchdog.action_shutdown.model_i6300esb.hotunplug_test: PASS (17.47 s)
 (04/31) type_specific.io-github-autotest-libvirt.watchdog.action_shutdown.model_ib700.model_test: PASS (48.42 s)
 (05/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.long_domain_name.model_i6300esb.model_test: PASS (67.54 s)
 (06/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.long_domain_name.model_i6300esb.hotplug_test: -^[[/^[[PASS (68.23 s)
 (07/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.long_domain_name.model_i6300esb.hotunplug_test: PASS (16.99 s)
 (08/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.normal_domain_name.model_i6300esb.model_test: PASS (71.31 s)
 (09/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.normal_domain_name.model_i6300esb.hotplug_test: PASS (67.76 s)
 (10/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.normal_domain_name.model_i6300esb.hotunplug_test: PASS (17.05 s)
 (11/31) type_specific.io-github-autotest-libvirt.watchdog.action_dump.normal_domain_name.model_ib700.model_test: /^[[/^[[PASS (67.45 s)
 (12/31) type_specific.io-github-autotest-libvirt.watchdog.action_none.model_i6300esb.model_test: \^[[\^[[PASS (197.54 s)
 (13/31) type_specific.io-github-autotest-libvirt.watchdog.action_none.model_i6300esb.hotplug_test: PASS (198.66 s)
 (14/31) type_specific.io-github-autotest-libvirt.watchdog.action_none.model_i6300esb.hotunplug_test: PASS (21.05 s)
 (15/31) type_specific.io-github-autotest-libvirt.watchdog.action_none.model_ib700.model_test: PASS (197.82 s)
 (16/31) type_specific.io-github-autotest-libvirt.watchdog.action_poweroff.model_i6300esb.model_test: |^[[/^[[\^[[PASS (47.67 s)
 (17/31) type_specific.io-github-autotest-libvirt.watchdog.action_poweroff.model_i6300esb.hotplug_test: -^[[PASS (49.45 s)
 (18/31) type_specific.io-github-autotest-libvirt.watchdog.action_poweroff.model_i6300esb.hotunplug_test: PASS (17.14 s)
 (19/31) type_specific.io-github-autotest-libvirt.watchdog.action_poweroff.model_ib700.model_test: PASS (47.62 s)
 (20/31) type_specific.io-github-autotest-libvirt.watchdog.action_pause.model_i6300esb.model_test: PASS (48.15 s)
 (21/31) type_specific.io-github-autotest-libvirt.watchdog.action_pause.model_i6300esb.hotplug_test: PASS (49.18 s)
 (22/31) type_specific.io-github-autotest-libvirt.watchdog.action_pause.model_i6300esb.hotunplug_test: PASS (17.13 s)
 (23/31) type_specific.io-github-autotest-libvirt.watchdog.action_pause.model_ib700.model_test: \^[[|^[[PASS (48.39 s)
 (24/31) type_specific.io-github-autotest-libvirt.watchdog.action_reset.model_i6300esb.model_test: PASS (65.44 s)
 (25/31) type_specific.io-github-autotest-libvirt.watchdog.action_reset.model_i6300esb.hotplug_test: PASS (65.72 s)
 (26/31) type_specific.io-github-autotest-libvirt.watchdog.action_reset.model_i6300esb.hotunplug_test: PASS (17.01 s)
 (27/31) type_specific.io-github-autotest-libvirt.watchdog.action_reset.model_ib700.model_test: PASS (68.89 s)
 (28/31) type_specific.io-github-autotest-libvirt.watchdog.action_inject_nmi.model_i6300esb.model_test: -^[[-^[[PASS (48.25 s)
 (29/31) type_specific.io-github-autotest-libvirt.watchdog.action_inject_nmi.model_i6300esb.hotplug_test: PASS (48.87 s)
 (30/31) type_specific.io-github-autotest-libvirt.watchdog.action_inject_nmi.model_i6300esb.hotunplug_test: PASS (17.00 s)
 (31/31) type_specific.io-github-autotest-libvirt.watchdog.action_inject_nmi.model_ib700.model_test: PASS (47.89 s)
RESULTS    : PASS 31 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1875.92 s

**Comment by iccaszhulili** [comment]
rh-caselink retest please

File: libvirt/tests/src/virtual_device/watchdog.py [hunk 1]
```diff
@@ -93,36 +95,52 @@ def _check_dump_file(dump_path, domain_id):
...
+            else:
+                logging.debug("Guest is not in shutoff state since watchdog action is none.")
         elif action == "inject-nmi":
             if not utils_misc.wait_for(_inject_nmi, 180, 10):
                 test.fail("Guest not receive inject-nmi after watchdog triggered\n")
             elif not utils_misc.wait_for(_inject_nmi_event, 180, 10):
                 test.fail("No inject-nmi watchdog event caught")
+            else:
+                logging.debug("Guest received inject-nmi and inject-nmi watchdog event "
+                              " has been caught.")
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_device/watchdog.py]: additional prefixing space can be removed.

File: libvirt/tests/src/virtual_device/watchdog.py [hunk 2]
```diff
@@ -146,11 +168,40 @@ def _check_dump_file(dump_path, domain_id):
...
+            vmxml.add_device(watchdog_dev)
+            vmxml.sync()
+            try:
+                vm.start()
+            except Exception:
+                test.fail("VM startup after adding watchdog device failed!")
+
+        elif hotplug_test:
+            watchdog_xml = os.path.join(data_dir.get_tmp_dir(), "tmp_watchdog_xml")
+            watchdog_xml = watchdog_dev.xml
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_device/watchdog.py]: this piece code is not making sense, `watchdog_xml` is assigned twice ?

File: libvirt/tests/src/virtual_device/watchdog.py [hunk 3]
```diff
@@ -146,11 +168,40 @@ def _check_dump_file(dump_path, domain_id):
...
+            attach_result = virsh.attach_device(vm_name, watchdog_xml,
+                                                ignore_status=True, debug=True)
         session = vm.wait_for_login()
+
+        # No need to trigger watchdog after hotunplug
+        if hotunplug_test:
+            cur_xml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+            cur_watchdog = cur_xml.xmltreefile.find('devices/watchdog')
+            cur_watchdog_xml = os.path.join(data_dir.get_tmp_dir(), "cur_watchdog_xml")
+            cur_watchdog_xml = Watchdog.new_from_element(cur_watchdog).xml
```
Comment (balamuruhans) [inline libvirt/tests/src/virtual_device/watchdog.py]: same here

File: libvirt/tests/src/virtual_device/watchdog.py (hunk 2 above)
Comment (iccaszhulili) [reply libvirt/tests/src/virtual_device/watchdog.py]: The code in first line is used to generate a file, the second file is used to write the xml of watchdog instance into that file

File: libvirt/tests/cfg/virtual_device/watchdog.cfg [hunk 4]
```diff
@@ -5,8 +5,18 @@
...
             model = "i6300esb"
+            variants:
+                - model_test:
+                    model_test = "yes"
+                - hotplug_test:
+                    hotplug_test = "yes"
+                - hotunplug_test:
+                    hotunplug_test = "yes"
         - model_ib700:
             model = "ib700"
```
Comment (kylazhang) [inline libvirt/tests/cfg/virtual_device/watchdog.cfg]: please add "no q35" for ib700 testing as it is not supported, thanks

**Comment by fangge1212** [comment]
Need to add pcie-to-pci-bridge controller for q35 vm, or hotplug i6300esb watchdog will fail.

**Comment by fangge1212** [comment]
Test result:
`2019-07-18 23:45:02,053 process          L0714 INFO | Command '/usr/bin/virsh attach-device avocado-vt-vm1 /var/tmp/xml_utils_temp_p_xv3fmp.xml' finished with 0 after 0.07289624214172363s
2019-07-18 23:45:02,067 virsh            L0706 DEBUG| status: 0
2019-07-18 23:45:02,068 virsh            L0707 DEBUG| stdout: Device attached successfully
2019-07-18 23:45:02,068 virsh            L0708 DEBUG| stderr:
2019-07-18 23:45:02,069 virt_vm          L1167 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2019-07-18 23:45:02,070 utils_misc       L0588 DEBUG| Retry verifying IP address (0.000014 secs)
2019-07-18 23:45:02,082 virt_vm          L0898 DEBUG| Found/Verified IP 192.168.122.21 for VM avocado-vt-vm1 NIC 0
2019-07-18 23:45:02,083 utils_misc       L0588 DEBUG| Retry verifying IP address (0.000018 secs)
2019-07-18 23:45:03,382 client           L1126 DEBUG| Sending command: gsettings set org.gnome.settings-daemon.plugins.power button-power shutdown
2019-07-18 23:45:03,498 client           L1126 DEBUG| Sending command: echo $?
2019-07-18 23:45:03,600 client           L1126 DEBUG| Sending command: dmesg | grep -i i6300esb && lsmod | grep i6300esb
2019-07-18 23:45:03,712 client           L1126 DEBUG| Sending command: echo $?
2019-07-18 23:45:03,938 libvirt_vm       L2140 DEBUG| Destroying VM
2019-07-18 23:45:04,302 libvirt_vm       L2175 WARNI| Requested MAC address release from persistent vm avocado-vt-vm1. Ignoring.
2019-07-18 23:45:04,433 virsh            L1434 DEBUG| Undefine VM avocado-vt-vm1
2019-07-18 23:45:04,464 virsh            L1418 DEBUG| Define VM from /var/tmp/xml_utils_temp_cbot4a3h.xml
2019-07-18 23:45:04,558 stacktrace       L0041 ERROR|
2019-07-18 23:45:04,558 stacktrace       L0044 ERROR| Reproduced traceback from: /var/tmp/libvirt-ci/runtest/avocado-vt/avocado-vt/avocado_vt/test.py:449
2019-07-18 23:45:04,562 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-07-18 23:45:04,562 stacktrace       L0047 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-libvirt/libvirt/tests/src/virtual_device/watchdog.py", line 55, in trigger_watchdog
2019-07-18 23:45:04,562 stacktrace       L0047 ERROR|     session.cmd("dmesg | grep -i %s && lsmod | grep %s" % (model, model))
2019-07-18 23:45:04,562 stacktrace       L0047 ERROR|   File "/root/.libvirt-ci-venv-ci-runtest-R8mwO1/lib/python3.6/site-packages/aexpect/client.py", line 1288, in cmd
2019-07-18 23:45:04,563 stacktrace       L0047 ERROR|     raise ShellCmdError(cmd, s, o)
2019-07-18 23:45:04,563 stacktrace       L0047 ERROR| aexpect.exceptions.ShellCmdError: Shell command failed: 'dmesg | grep -i i6300esb && lsmod | grep i6300esb'    (status: 1,    output: '')
`

**Comment by fangge1212** [comment]
> dmesg | grep -i i6300esb && lsmod | grep i6300esb

Maybe a short sleep is needed after attach-device in case watchdog device didn't showup in vm immediatelly

**Comment by fangge1212** [comment]
And also a short sleep is needed after detach-device in case watchdog device didn't disappear in vm immediatelly

**Comment by fangge1212** [comment]
I submitted a new pr based on this pr: https://github.com/autotest/tp-libvirt/pull/2244

**Comment by iccaszhulili** [comment]
has been covered in https://github.com/autotest/tp-libvirt/pull/2244

**Pull request closed without merging**